File: packages/merge/src/source-reading/merge-source.contract.ts
import type { Raster } from '@scanmate/ink'

/**
 * Anything that can become pages of the merged PDF.
 *
 * - a path (`string` or file `URL`) or bytes (`Uint8Array`, `ArrayBuffer`) -
 *   a PDF or any image libvips reads, told apart by content, not by name;
 * - a decoded {@link Raster};
 * - an image with its resolution, `{ raster, dpi }` - which is what every
 *   `PageImage` in the pipeline is, so an aligned or enhanced page goes in as it
 *   is. When it also carries its encoded `image`, those bytes are embedded
 *   rather than encoding the raster again.
 */

/** What a source turned out to be. */
export type SourceKind = 'pdf' | 'image' | 'raster'

/** A source that could not be used, and which one it was. */
export class MergeSourceError extends Error {
  /**
   * @param index - Zero-based position of the source in the list given.
   * @param message - What is wrong with it.
   */
  constructor (readonly index: number, message: string) {
    super(`source ${index + 1}: ${message}`)
    this.name = 'MergeSourceError'
  }
}

File: python-packages/scanmate-merge/scanmate_merge/source_reading/merge_source_contract.py
"""What a source is, and what it means for one to be unusable."""

from __future__ import annotations

from typing import Literal, TypeAlias

#: What a source turned out to be.
SourceKind: TypeAlias = Literal["pdf", "image", "raster"]


class MergeSourceError(Exception):
    """A source that could not be used, and which one it was."""

    def __init__(self, index: int, message: str) -> None:
        """Name the source by its position, as a caller counts them.

        :param index: Zero-based position of the source in the list given.
        :param message: What is wrong with it.
        """
        super().__init__(f"source {index + 1}: {message}")
        #: Zero-based position of the source in the list given.
        self.index = index
